cabal-install/src/Distribution/Client/ProjectPlanning.hs

```diff
@@ -2361,35 +2369,114 @@ elaborateInstallPlan
+        -- If we don't know for certain, just assume we can
+        -- which matches behaviour in previous cabal releases
+        Nothing -> True
+
+      canBuildSharedLibs = canBuildWayLibs dynamicSupported
+      canBuildProfilingLibs = canBuildWayLibs profilingVanillaSupported
+      canBuildProfilingSharedLibs = canBuildWayLibs profilingDynamicSupported
+
+      wayWarnings pkg = do
+        when
+          (needsProfilingLib pkg && not canBuildProfilingLibs)
+          (warnProgress (text "Compiler does not support building p libraries, profiling is disabled"))
+        when
+          (needsSharedLib pkg && not canBuildSharedLibs)
+          (warnProgress (text "Compiler does not support building dyn libraries, dynamic libraries are disabled"))
+        when
+          (needsProfilingLibShared pkg && not canBuildProfilingSharedLibs)
+          (warnProgress (text "Compiler does not support building p_dyn libraries, profiling dynamic libraries are disabled."))
+
       pkgsUseProfilingLibrary :: Set PackageId
       pkgsUseProfilingLibrary =
         packagesWithLibDepsDownwardClosedProperty needsProfilingLib
+
+      needsProfilingLib pkg =
+        fromFlagOrDefault compilerShouldUseProfilingLibByDefault (profBothFlag <> profLibFlag)
         where
-          needsProfilingLib pkg =
-            fromFlagOrDefault False (profBothFlag <> profLibFlag)
-            where
-              pkgid = packageId pkg
-              profBothFlag = lookupPerPkgOption pkgid packageConfigProf
-              profLibFlag = lookupPerPkgOption pkgid packageConfigProfLib
+          pkgid = packageId pkg
+          profBothFlag = lookupPerPkgOption pkgid packageConfigProf
+          profLibFlag = lookupPerPkgOption pkgid packageConfigProfLib
+
+      pkgsUseProfilingLibraryShared :: Set PackageId
+      pkgsUseProfilingLibraryShared =
+        packagesWithLibDepsDownwardClosedProperty needsProfilingLibShared
+
+      needsProfilingLibShared pkg =
+        fromMaybe
+          compilerShouldUseProfilingSharedLibByDefault
+          -- Case 1: If --enable-profiling-shared is passed explicitly, honour that
+          ( case profLibSharedFlag of
+              Just v -> Just v
+              Nothing -> case pkgDynExe of
+                -- case 2: If --enable-executable-dynamic is passed then turn on
+                -- shared library generation.
+                Just True ->
+                  -- Case 3: If --enable-profiling is passed, then we are going to
+                  -- build profiled dynamic, so no need for shared libraries.
+                  case pkgProf of
+                    Just True -> if canBuildProfilingSharedLibs then Just True else Nothing
+                    _ -> Nothing
+                -- But don't necessarily turn off shared library generation is
+                -- --disable-executable-dynamic is passed. The shared objects might
+                -- be needed for something different.
+                _ -> Nothing
+          )
```

Comment:
Just changing the comments (stale copy pasta):

```suggestion
          ( case profLibSharedFlag of
              Just v -> Just v
              Nothing -> case pkgDynExe of
                Just True ->
                  case pkgProf of
                   -- case 2: --enable-executable-dynamic + --enable-profiling
                   -- turn on shared profiling libraries
                    Just True -> if canBuildProfilingSharedLibs then Just True else Nothing
                    _ -> Nothing
                _ -> Nothing
          )
```

However, shouldn't we be returning `Just False` in some cases? The comment for not returning `Just False` for `needsSharedLib` don't seem to apply here; for example if `canBuildProfilingSharedLibs` is `False` I would expect to get `Just False` here (although that may be what we get in practice due to the value of `compilerShouldUseProfilingSharedLibByDefault`).

Reply:
This part of the code is about determining what the user is requesting and how certain they are of that. A `Just` value indicates that we are certain that we require something. `Just True` => certainly build prof shared libs, `Just False` => certainly don't build prof shared libs. If we are unsure then it falls back to the default.

Once we have established what the user has requested, then this is matched up against what the compiler provides (see warnings which check `needsProfilingLibShared` against `canBuildProfilingSharedLibs`) and also `withProfLibShared`, which is only enabled when the compiler supports that.

`Just False` is returned only in the case when you pass `--disable-profiling-shared` as `Just False` indicates that we definitely do not want to build these libraries.